### PR TITLE
proposing additions for type annotation

### PR DESCRIPTION
**Pull request by smoothdeveloper**
Proposing few examples on how to use whitespace for type annotations.

### CODE REVIEW

File: docs/fsharp/style-guide/formatting.md [hunk 1]
```diff
@@ -82,6 +82,46 @@ type MyFun = int -> int -> string
...
+
+```fsharp
+// OK
+let myFun (a: decimal) b c = a + b + c
+
+// Bad
+let myFunBad (a:decimal)(b)c = a + b + c
+```
+
+### type annotation
```
Comment (cartermp) [inline docs/fsharp/style-guide/formatting.md]:
```suggestion
### Type annotations
```

File: docs/fsharp/style-guide/formatting.md [hunk 2]
```diff
@@ -82,6 +82,46 @@ type MyFun = int -> int -> string
...
+
+#### Right-pad function argument type annotation with white space
```
Comment (cartermp) [inline docs/fsharp/style-guide/formatting.md]:
```suggestion
#### Right-pad function argument type annotations
```

File: docs/fsharp/style-guide/formatting.md [hunk 3]
```diff
@@ -82,6 +82,46 @@ type MyFun = int -> int -> string
...
+
+When defining arguments with type annotation, use white space after the `:` symbol:
```
Comment (cartermp) [inline docs/fsharp/style-guide/formatting.md]:
```suggestion
When defining arguments with type annotations, use white space after the `:` symbol:
```

File: docs/fsharp/style-guide/formatting.md [hunk 4]
```diff
@@ -82,6 +82,46 @@ type MyFun = int -> int -> string
...
+
+```fsharp
+// OK
+let complexFunction (a: int) (b: int) c = a + b + c
+
+// Bad
+let complexFunctionBad (a :int) (b :int) (c:int) = a + b + c
+```
+
+#### Surround definition type annotations with white space
```
Comment (cartermp) [inline docs/fsharp/style-guide/formatting.md]:
```suggestion
#### Surround return type annotations with white space
```

File: docs/fsharp/style-guide/formatting.md [hunk 5]
```diff
@@ -82,6 +82,46 @@ type MyFun = int -> int -> string
...
+
+In a let binding with type annotation, use white space after the `:` symbol:
```
Comment (cartermp) [inline docs/fsharp/style-guide/formatting.md]:
```suggestion
In a let-bound function or value type annotation (return type in the case of a function), use white space before and after the `:` symbol:
```

File: docs/fsharp/style-guide/formatting.md [hunk 6]
```diff
@@ -82,6 +82,46 @@ type MyFun = int -> int -> string
...
+
+```fsharp
+// OK
+let expensiveToCompute : int = 0
```
Comment (cartermp) [inline docs/fsharp/style-guide/formatting.md]:
```suggestion
let expensiveToCompute : int = 0 // Type annotation for let-bound value
```

File: docs/fsharp/style-guide/formatting.md [hunk 7]
```diff
@@ -82,6 +82,46 @@ type MyFun = int -> int -> string
...
+let myFun (a: decimal) b c : decimal = a + b + c
```
Comment (cartermp) [inline docs/fsharp/style-guide/formatting.md]:
```suggestion
let myFun (a: decimal) b c : decimal = a + b + c // Type annotation for the return type of a function
```

**Review by cartermp — CHANGES_REQUESTED** [review]
Some minor changes, but otherwise this is great. Thanks @smoothdeveloper

**Review by cartermp — APPROVED** [review]

**Comment by auduchinok** [comment]
I've seen a lot of code with no space before colon in both values and function annotations, like this:
```fsharp
// values:

let x: SomeType = y

let f () =
    let y: AnotherType = g x


// functions:

let f (arg: _): unit =
    ()

let f x: unit =
    ()
```

Perhaps, I've seen more code formatted like this than with space like `let a : int = 1`.

@cartermp @dsyme I think we should explicitly allow this style, since Fantomas uses these rules to enforce arguably less popular code style without any option to keep no-space style. What do you think?

**Comment by dsyme** [comment]
I personally regard the former style (no space) as standard with the exception of having a whitespace here, partly so `):` doesn't look like a weird sad backward smilely

```fsharp
let f (arg: _) : unit =
    ()
```

We should document this here, yes

**Comment by dsyme** [comment]
@auduchinok I made a proposed adjustment here: https://github.com/dotnet/docs/pull/24643

**Pull request merged**